### ISSUE TITLE
Iterating a \Google\Protobuf\Internal\Message PHP object segfaults when using the protobuf extension

### ISSUE THREAD

**Issue by mackenziestarr** [protocolbuffers/protobuf#7319]
**What version of protobuf and what language are you using?**
Version: v3.11.4
Language: PHP

**What operating system (Linux, Windows, ...) and version?**
Linux, CentOS Linux release 7.5.1804 

**What runtime / compiler are you using (e.g., python version or gcc version)**
PHP 7.3.10 with protobuf extension https://pecl.php.net/package/protobuf/3.11.4

**What did you do?**
1. Define a protobuf message:
```proto3
syntax = "proto3";                                                                                                                                                                                           
package example;                                                                                                                                                                                             
                                                                                                                                                                                                             
message ProtoMessage {                                                                                                                                                                                       
  string msg = 1;                                                                                                                                                                                            
}
```
2. Compile with protoc
```                                                                                                                                                                                                            
-bash-4.2$ /usr/local/bin/protoc  --version
libprotoc 3.11.4
-bash-4.2$ /usr/local/bin/protoc  --proto_path=src --php_out=build/gen src/example.proto 
-bash-4.2$ tree build/gen/
build/gen/
├── Example
│   └── ProtoMessage.php
└── GPBMetadata
    └── Example.php
```         
3. Run this PHP script
```php
<?php                                                                                                                                                                                                        
require 'build/gen/GPBMetadata/Example.php';                                                                                                                                                                 
require 'build/gen/Example/ProtoMessage.php';                                                                                                                                                                                                                                                                                                                                                                             
echo "iterating generated php class ProtoMessage\n";                                                                                                                                                         
$proto_msg = new \Example\ProtoMessage();                                                                                                                                                                    
foreach ($proto_msg as $key => $value) {                                                                                                                                                                     
    assert(false);                                                                                                                                                                                           
} 
```

**What did you expect to see**

I expect the process to not segfault

**What did you see instead?**

The process segfaults
```
-bash-4.2$ php reproduce-segfault.php 
iterating generated php class ProtoMessage
Magick: abort due to signal 11 (SIGSEGV) "Segmentation Fault"...
Aborted
```

gdb backtrace
```
Program received signal SIGSEGV, Segmentation fault.
zend_hash_iterator_add (ht=0x0, pos=pos@entry=0) at /usr/src/debug/php-src-php-7.3.10/Zend/zend_hash.c:404
404             if (EXPECTED(!HT_ITERATORS_OVERFLOW(ht))) {
(gdb) bt
#0  0x000000000088058e in zend_hash_iterator_add (ht=0x0, pos=pos@entry=0) at /usr/src/debug/php-src-php-7.3.10/Zend/zend_hash.c:404
#1  0x00000000008e06b6 in ZEND_FE_RESET_R_SPEC_CV_HANDLER () at /usr/src/debug/php-src-php-7.3.10/Zend/zend_vm_execute.h:37513
#2  0x00000000008fa3cf in execute_ex (ex=0x0) at /usr/src/debug/php-src-php-7.3.10/Zend/zend_vm_execute.h:59321
#3  0x000000000090162c in zend_execute (op_array=op_array@entry=0x7fffeb4872a0, return_value=return_value@entry=0x0) at /usr/src/debug/php-src-php-7.3.10/Zend/zend_vm_execute.h:60889
#4  0x0000000000873ece in zend_execute_scripts (type=-348008400, type@entry=8, retval=retval@entry=0x0, file_count=file_count@entry=3) at /usr/src/debug/php-src-php-7.3.10/Zend/zend.c:1568
#5  0x00000000008100e0 in php_execute_script (primary_file=primary_file@entry=0x7fffffffc210) at /usr/src/debug/php-src-php-7.3.10/main/main.c:2639
#6  0x0000000000903b96 in do_cli (argc=2, argv=0x1447490) at /usr/src/debug/php-src-php-7.3.10/sapi/cli/php_cli.c:997
#7  0x0000000000466128 in main (argc=2, argv=0x1447490) at /usr/src/debug/php-src-php-7.3.10/sapi/cli/php_cli.c:1389
```

**Comment by Alex-David-Hu** [comment]
"Segmentation Fault".




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Mackenzie Starr"<notifications@github.com&gt;;
发送时间:&nbsp;2020年3月20日(星期五) 上午6:17
收件人:&nbsp;"protocolbuffers/protobuf"<protobuf@noreply.github.com&gt;;
抄送:&nbsp;"Subscribed"<subscribed@noreply.github.com&gt;;
主题:&nbsp;[protocolbuffers/protobuf] Iterating a \Google\Protobuf\Internal\Message PHP object segfaults when using the protobuf extension (#7319)





 
What version of protobuf and what language are you using?
 Version: v3.11.4
 Language: PHP
 
What operating system (Linux, Windows, ...) and version?
 Linux, CentOS Linux release 7.5.1804
 
What runtime / compiler are you using (e.g., python version or gcc version)
 PHP 7.3.10 with protobuf extension https://pecl.php.net/package/protobuf/3.11.4
 
What did you do?
 Steps to reproduce the behavior:
  
Define a protobuf message:
  syntax = "proto3";                                                                                                                                                                                            package example;                                                                                                                                                                                                                                                                                                                                                                                                            message ProtoMessage {                                                                                                                                                                                          string msg = 1;                                                                                                                                                                                             }   
Compile with protoc
  -bash-4.2$ /usr/local/bin/protoc  --version libprotoc 3.11.4 -bash-4.2$ /usr/local/bin/protoc  --proto_path=src --php_out=build/gen src/example.proto  -bash-4.2$ tree build/gen/ build/gen/ ├── Example │&nbsp;&nbsp; └── ProtoMessage.php └── GPBMetadata     └── Example.php   
Run this PHP script
  <?php                                                                                                                                                                                                         require 'build/gen/GPBMetadata/Example.php';                                                                                                                                                                  require 'build/gen/Example/ProtoMessage.php';                                                                                                                                                                                                                                                                                                                                                                              echo "iterating generated php class ProtoMessage\n";                                                                                                                                                          $proto_msg = new \Example\ProtoMessage();                                                                                                                                                                     foreach ($proto_msg as $key =&gt; $value) {                                                                                                                                                                          assert(false);                                                                                                                                                                                            } 
 
What did you expect to see
 
I expect the process to not segfault
 
What did you see instead?
 
The process segfaults
 -bash-4.2$ php reproduce-segfault.php  iterating generated php class ProtoMessage Magick: abort due to signal 11 (SIGSEGV) "Segmentation Fault"... Aborted  
gdb backtrace
 (gdb) bt #0  0x000000000088058e in zend_hash_iterator_add (ht=0x0, pos=pos@entry=0) at /usr/src/debug/php-src-php-7.3.10/Zend/zend_hash.c:404 #1  0x00000000008e06b6 in ZEND_FE_RESET_R_SPEC_CV_HANDLER () at /usr/src/debug/php-src-php-7.3.10/Zend/zend_vm_execute.h:37513 #2  0x00000000008fa3cf in execute_ex (ex=0x0) at /usr/src/debug/php-src-php-7.3.10/Zend/zend_vm_execute.h:59321 #3  0x000000000090162c in zend_execute (op_array=op_array@entry=0x7fffeb4872a0, return_value=return_value@entry=0x0) at /usr/src/debug/php-src-php-7.3.10/Zend/zend_vm_execute.h:60889 #4  0x0000000000873ece in zend_execute_scripts (type=-348008400, type@entry=8, retval=retval@entry=0x0, file_count=file_count@entry=3) at /usr/src/debug/php-src-php-7.3.10/Zend/zend.c:1568 #5  0x00000000008100e0 in php_execute_script (primary_file=primary_file@entry=0x7fffffffc210) at /usr/src/debug/php-src-php-7.3.10/main/main.c:2639 #6  0x0000000000903b96 in do_cli (argc=2, argv=0x1447490) at /usr/src/debug/php-src-php-7.3.10/sapi/cli/php_cli.c:997 #7  0x0000000000466128 in main (argc=2, argv=0x1447490) at /usr/src/debug/php-src-php-7.3.10/sapi/cli/php_cli.c:1389  
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by smalon** [comment]
i got the same problem here

**Comment by jbq** [comment]
Happened to me as well, crashed in production, here is a gdb trace: https://gist.github.com/jbq/38d958924bd0205bb50fe0be1005b09b

**Comment by adsr** [comment]
This should be fixed in PHP 7.4.12+.

[0] https://bugs.php.net/bug.php?id=80186
[1] https://github.com/php/php-src/pull/6270
[2] https://github.com/php/php-src/commit/15443f8af4783ce6ea5362ec5a38a62b69ff8447

**Comment by haberman** [comment]
Thanks @adsr, it looks like this bug was fixed in PHP itself.
